src/_common/video/video.ts

```diff
@@ -81,4 +124,137 @@ export default class AppVideo extends Vue {
 			this.video.load();
 		}
 	}
+
+	private trackPlaytime() {
+		// Only track playtime events for components that get a VideoPlayerController passed in (meaning that we have playback controls).
+		if (!this.videoStartTime || !this.inheritedPlayer) {
+			return;
+		}
+
+		const playtime = Date.now() - this.videoStartTime;
+		const loops = Math.floor(playtime / this.player.duration);
+		trackVideoPlayerEvent(this.player, 'watched', 'playtime', `${Math.ceil(playtime / 1000)}`);
+		trackVideoPlayerEvent(this.player, 'watched', 'loops', `${loops}`);
+		this.videoStartTime = 0;
+	}
+
+	private setupVideoEvents() {
+		if (!this.video || !this.player) {
+			return;
+		}
+		this.video.addEventListener('play', () => {
+			this.player.state = 'playing';
+			this.videoStartTime = Date.now();
+		});
+		this.video.addEventListener('pause', () => {
+			this.player.state = 'paused';
+			this.trackPlaytime();
+		});
+		this.video.addEventListener('volumechange', () => (this.player.volume = this.video.volume));
+		this.video.addEventListener('durationchange', () => {
+			if (this.video.duration) {
+				this.player.duration = this.video.duration * 1000;
+			}
+		});
+		this.video.addEventListener(
+			'timeupdate',
+			() => (this.player.currentTime = this.video.currentTime * 1000)
+		);
+		this.video.addEventListener('progress', () => {
+			let time = 0;
+			for (let i = 0; i < this.video.buffered.length; ++i) {
+				time = Math.max(time, this.video.buffered.end(i) * 1000);
+			}
+			this.player.bufferedTo = time;
+		});
+	}
+
+	private async tryPlayingVideo() {
+		if (!this.video || !this.player) {
+			return;
+		}
+
+		const startVolume = this.player.volume;
+		try {
+			await this.video.play();
+			return;
+		} catch {}
+
+		if (!this.allowDegradedAutoplay) {
+			this.player.state = 'paused';
+			return;
+		}
+
+		// If autoplaying the video failed, first try setting the volume of
+		// the video to 0.
+		if (startVolume > 0) {
+			this.player.volume = 0;
+			await this.$nextTick();
+		}
+
+		try {
+			await this.video.play();
+			return;
+		} catch {}
+
+		// If the autoplaying is still blocked with the volume set to 0,
+		// pause the video in the player controller and reset the player
+		// volume to the initial setting.
+		this.player.state = 'paused';
+		this.player.volume = startVolume;
+	}
+
+	@Watch('shouldPlay')
+	onShouldPlayChange() {
+		// Ignore 'shouldPlay' changes for instances that have an inherited player controller handling playback.
+		if (this.inheritedPlayer) {
```

Comment:
I don't think we need this check. shouldPlay will never change anyway unless they bind to it

Reply:
removed